### DISCUSSION TITLE
Access to system audio/microphone settings within Jamulus GUI

### DISCUSSION THREAD

**Discussion by henkdegroot** [jamulussoftware/jamulus#2181] [Ideas]
I have noticed that many people, using Jamulus on a Windows platform, have some difficulties during the initial setup to set the windows microphone level correct when you are not using an audio device.
One of the main problems is how to get to that function can be a bit hard.
Also, once set, people tent to forget how they got there and do not have an easy way to get back into that window.

I.e. the window I aim at is this:
![sounds-settings-recording](https://user-images.githubusercontent.com/13550012/147328417-998e0cbd-3c35-4791-a049-1e60bacf895e.PNG)

(Sorry...only have a Dutch screen available)

I believe it would enhance the software functionality to have a button in the Audio-/Netwerk-settings window to open this functionality. Perhaps even adding two buttons to this page, one for input and one for output.

Example of this could be:
![NewButton](https://user-images.githubusercontent.com/13550012/147329048-b9df0038-e058-44d7-8eff-ebc146189fe0.png)

(In the example only a button for Microphone, but it would be possible to add one for output as well though the output is easier accessible in windows and might be a bit overkill).

Obvious, this change would only impact the Windows version.

**Comment by ann0see** [comment]
Yeah, I think there already was a discussion about that somewhere. Ideally, there would be a sync with the mic gain feature in the advanced settings. This would surely require some larger changes - and maybe the use of the ASIO4ALL API.

**Reply by henkdegroot** [reply]
I don't think that the mic gain feature has any influence on the windows gain settings for the microphone.
Adding the button would be an easy item to do.  Already got it working on a local build :) 

Just want to see if there is interest in this....

**Reply by dcorson-ticino-com** [reply]
I think this would really make things easier for many Windows users (and MAC users).
In my environment there are almost no Win10 users, but many of the MAC users have had problems finding the mic setting too (and also the permission to use the mic for Jamulus).
I agree that this has nothing to do with the mic gain in the settings window.  I have not yet seen anyone who needed to use that.

**Comment by chrisrimple** [comment]
Maybe name the button "System Microphone Settings", so that on Mac it could redirect to System Preferences > Sound > Input?

**Reply by henkdegroot** [reply]
Hmm...nice, while I know how to get this done on Windows...I don't know how to program that on macOS. Interesting suggestion.

**Reply by henkdegroot** [reply]
@chrisrimple @dcorson-ticino-com 
I found a method which allows me to open the System Preference Sound settings. It just seem to be really complex (or at least for me) to get that opening the Input tab direct. 

Guess it will be enhancement for both OS to get quick to the Sound Settings and the button label can just be "System Sound Settings".
Thinking it will be a lot easier to tell someone to click that button and then select the Input tab.

**Reply by pljones** [reply]
Is there anything similar on Linux?  Android?  iOS?

**Comment by ann0see** [comment]
@henkdegroot decided not to implement this button. Maybe another way to easily open system sound settings could be implemented (without causing too much confusion). 

Maybe a Menu entry?

**Reply by henkdegroot** [reply]
> without causing too much confusion

Correct, seems that adding a button as shown in the sample, could cause confusion and users select the wrong thing trying to fix something.
Any suggestions to a better solution are welcome.

**Reply by ann0see** [reply]
The main issue is mic gain control probably. Jamulus should be able to control this in e.g The advanced settings like the amplification.

**Reply by henkdegroot** [reply]
> Jamulus should be able to control this

That would be even better, hoping it will then also save the setting so that every time you use/start Jamulus it gets sets to the corresponding value set.

**Reply by pljones** [reply]
Dirty hacks like digitally amplifying a signal with a large amount of unused capacity isn't a great idea...

Ideally you want a user to set their system up properly so that they have good signal levels, no clipping and low noise.  All of that is _outside_ what Jamulus does.  Jamulus isn't for controlling people's sound set up.

Keep Jamulus focussed on the the problem it's designed to solve: online jamming -- getting your audio to a server for mixing and back.

Keep everything else outside.  Put it in the set up instructions or online guides.  But keep Jamulus simple.

**Comment by chrisrimple** [comment]
Not sure that I agree with @pljones.  There are many audio applications that provide access to system audio settings, specifically because system audio settings will impact the application.  While a button that opens system audio settings may not be the right answer, the need remains and is valid, particularly when we consider that Jamulus is being used by plenty of non-technical people (musicians, choral singers, etc.).  

As an alternative to a button, what about a slider for mic gain control?  Moving it adjusts the system settings.  MANY audio apps have this feature, not unlike many playback apps having a volume control slider that adjusts system volume (v. application volume).

**Reply by pljones** [reply]
> There are many audio applications that provide access to system audio settings, specifically because system audio settings will impact the application.

As far as I'm aware:
Reaper doesn't.
MuLab doesn't.
foobar2000 doesn't.
MPC-HC doesn't.
Audacity doesn't.  (Although its built in faders seem to adjust whatever underlying audio system you're using.  PortAudio seems to do that without letting you see what's happening.)

I don't think I use any software that does, to be honest.  Maybe some does...

For ASIO, _except when some ASIO-emulation is in use_, the system sound controls are almost always meaningless.  If you've decided to use some ASIO-emulation, then I'd expect you to use it to control the system sound controls and, if it doesn't allow that, to get it changed -- rather than thinking it's a Jamulus problem that _only_ happens because you've chosen to use some ASIO-emulation.

**Reply by rdica** [reply]
@chrisrimple How would this work for Linux users?

**Reply by chrisrimple** [reply]
No idea, not sure that it's an option on Linux.  For an OS where the system input can't be controlled, disable the control.

**Comment by gene96817** [comment]
Would the way Zoom helps with audio setup be a good example? See zoom.us/Preferences/Audio.
I'm a Mac user. I hope the Window's Zoom UI is the same.

**Reply by ann0see** [reply]
Yes, I think it works. I think it should be possible to control Windows audio somehow. Hopefully there’s an ASIO compatible way to do that.

**Reply by pljones** [reply]
> Yes, I think it works. I think it should be possible to control Windows audio somehow. Hopefully there’s an ASIO compatible way to do that.

The "ASIO Device Settings" should be all that's needed when using ASIO.  If you need more (e.g. the control panel for a specific device), you'll need a long list of all the possible hardware suppliers and where a user might have installed the control panel...
